semgrep-core/src/engine/Match_extract_mode.ml

```diff
@@ -386,22 +407,35 @@ let extract_as_separate erule_table xtarget rule_ids matches =
                    None
              in
              (* Read the extracted text from the source file *)
-             let contents =
+             let contents_raw =
                Common.with_open_infile m.file (fun chan ->
                    let extract_size = end_extract_pos - start_extract_pos in
                    seek_in chan start_extract_pos;
                    really_input_string chan extract_size)
              in
+             (* Convert from JSON to plaintext, if required *)
+             let contents =
+               let (`Extract { Rule.json; _ }) = erule.mode in
+               if json then extract_code_from_json contents_raw
+               else contents_raw
+             in
              logger#trace
                "Extract rule %s extracted the following from %s at bytes %d-%d\n\
                 %s"
                m.rule_id.id m.file start_extract_pos end_extract_pos contents;
              (* Write out the extracted text in a tmpfile *)
              let (`Extract { Rule.dst_lang; _ }) = erule.mode in
              let target = mk_extract_target dst_lang contents rule_ids in
+             (* For some reason, with json extraction option, it needs a fix to point the right line *)
+             (* TODO: Find the reason of this behaviour and fix it properly *)
              let map_loc =
-               map_loc start_extract_pos line_offset col_offset
-                 xtarget.Xtarget.file
+               let (`Extract { Rule.json; _ }) = erule.mode in
```

Comment:
We already have to get some fields out of the rule a couple lines prior, so maybe something like this instead?
```ocaml
let (`Extract { Rule.dst_lang; json; _ }) = erule.mode in (* << just getting both members here *)
```
and then delete new line 432.

Reply:
Done.